### DISCUSSION TITLE
EdgeTX Presets - a new tool

### DISCUSSION THREAD

**Discussion by offer-shmuely** [EdgeTX/edgetx#3689] [Ideas]
I like promote a new concept (to edgeTX) called **_Presets_**.
The idea is taken from betaflight 4.3, an so the name.

What are EdgeTX Presets?
**_EdgeTX Presets are basically little snippets of lua scripts, that you can apply to your current loaded model._** 

sample for presets:
- add motor safety/arm switch
- add dual-rate (actually 3 rates) for aileron/elevator
- load picture (yes, we have it, but we can do that on a script while showing the pictures itself, not just the name)
-  ??

once a preset is selected, it will show an explanation what it is going to do, ask the user for parameters (same as the wizard), and apply them to the current loaded model.

what do you way, is it worth the effort?

I think that first version can be implemented totally in lua, so it is not take resources if not activated, and it will not put the os in any risk during flight.
 
comments are welcome....

**Comment by pfeerick** [comment]
*Is it worth the effort?* Yes, I think so... makes it easier to get people to use some of the safety stuff they should be using, or to add otherwise complex logic or repetitive options to their models. 

I know Raphael had been thinking of a template concept whereby you could have snippets that were shared between models (e.g. maybe some arm switch logic) which would allow for an "edit once, use everywhere" concept (I guess you could liken it to global functions in some ways), but until that sees the light of day (if it does) this would be the next best thing. 

I think it would be good for the presets/snippets to be able to be added or removed (i.e. making them more like an option you can just toggle on or off at will), but I'm not sure that the removal part will be so easy to implement, so maybe it will have to be just added, and manual cleanup if it needs to be undone.

**Comment by offer-shmuely** [comment]
Do you have additional suggestions for future "presets"?

**Reply by offer-shmuely** [reply]
here is some sample from my last weekend work:

selecting is done currently from TOOLS (maybe in the future it will be inside the "manage models"?)
![image](https://github.com/EdgeTX/edgetx/assets/7026911/b03c1b33-5139-4d4c-adea-44a5bba13167)

can select a "preset" from existing one (ugly, but working)
![image](https://github.com/EdgeTX/edgetx/assets/7026911/da6e4eef-6c09-46f7-a97c-0887f6da3897)

you get some info from a meta.in fie (name/author/about/ver)
![image](https://github.com/EdgeTX/edgetx/assets/7026911/825ec55f-8ff6-4a28-8657-8f105c8ae2c9)

this is a per-preset page, here the writer of the preset write it's code
![image](https://github.com/EdgeTX/edgetx/assets/7026911/400e3102-0be7-4cb3-9709-543693ecda2d)

confirmation page
![image](https://github.com/EdgeTX/edgetx/assets/7026911/81e2956b-ce08-4114-b131-7b2e9f834f29)

![image](https://github.com/EdgeTX/edgetx/assets/7026911/678955cf-2e56-4ca4-84b5-3d27b8998d24)

**Reply by arkarkark** [reply]
> Do you have additional suggestions for future "presets"?

* This [throttle cut](https://youtu.be/Y0eKFAFNeX8) is complicated but awesome.
* 3 setting throws/expo there's regular, full throws and then in the middle the expo and throw is set by the potentiometers. with a button to press to read out the values. I also have a trigger switch to put it in full expo so I can do quick snap rolls.
* a telemetry based  low battery alarm and a button to press to read out battery level.
* a buddybox/trainer mode where the control is given over by pressing a momentary button, release and you have control. I also don't give over throttle control. Mine also says something when I give or take back control.
* powered plane flight countdown time that only increases when the non throttle locked throttle is > 0. with button to read out remaining time.
* glider plane flight time that only increases when the servos are moving. with button to read out time.

I have a Jumper T-Pro which has a lot of buttons and the trigger momentary press switches too.

I tried writing some snippets to set up some of these and I'm very interested in how you are going to do it. Anything that needs [logical switches](https://luadoc.edgetx.org/part_iii_-_opentx_lua_api_reference/model-functions-less-than-greater-than-luadoc-begin-model/setlogicalswitch) or special functions seems impossible right now since all the settings are made by indexing into the list of options and that list changes depending on many other settings. So trying to work out where SWA↓ would be in the list is pretty much impossible.

I think I could maybe get most of the benefit of this although not as nicely if I was able to copy/paste some or all of the logical switches or special functions pages between models.

**Reply by offer-shmuely** [reply]
It is complicated, since the api is not so clear, we need a lot of trial and error to achieve that goal, but i managed to search for "ail"&"ele" input, and insert there the dual rates

I managed also to create a special-function for the throttle lock (it also in the up coming new wizard) but i need to search for a free SF, and also test if line like this already exist

**Reply by arkarkark** [reply]
Thanks, today I found out about [getSwitchIndex](https://luadoc.edgetx.org/part_iii_-_opentx_lua_api_reference/general-functions-less-than-greater-than-luadoc-begin-general/getswitchindex-positionname) which I think is really going to help me. I'll probably be writing my code to do some of these presets I mentioned, but I'll likely do it in fixed reserved switch/function slots.

**Reply by offer-shmuely** [reply]
And what will happen on a model that this slot is already occupied?

**Reply by arkarkark** [reply]
I'm lucky I know all my models, but I could always error out with a warning, or overwrite them which would help with upgrades. My plan is to pretty much automate the generation of all my models. So I can run it again and again and overwrite old switches/functions with the new hotness. But maybe with these new lua methods I just learned about making it use the next available slots would be easy enough. It's going to be a little while for me, I have to move over from openTX to edgeTX now 😸

**Reply by Shonkonray** [reply]
@arkarkark
Maybe you could try and change the settings in the model.yml files. Those are quite clear and easy to read, but please just take it with a grain of salt, because i have absolutely no clue with text based programing (i am only good with plc FBD programming). If i wanted to copy or reposition many logical switches i would probably try it manually within the .yml files (copy, paste, search and replace and so on).

Also, if you like the motor start from "RC video reviews" you might like the [one](https://github.com/EdgeTX/edgetx/discussions/3801) i use.

**Reply by arkarkark** [reply]
Indeed, modifying the yaml files is the way I've decided to go. This way I get to do it in python and I can keep some extra metadata around if I need it. thanks for sharing your throttle system, that's a neat idea.

**Comment by offer-shmuely** [comment]
folder structure look like this:
![image](https://github.com/EdgeTX/edgetx/assets/7026911/af50602a-ebd7-42e6-9f8c-2c5168931d5e)

meta.ini 
![image](https://github.com/EdgeTX/edgetx/assets/7026911/94042cda-ce9e-4e15-b3b2-9502d16cb555)

**Comment by jmxp69** [comment]
I like the sound of this. I spend a lot of time working out new logic gizmos for EdgeTX and sometimes I figure something out I want on every model. Usually this involves a drink and a session with Companion...

**Comment by offer-shmuely** [comment]
some sample


![image](https://github.com/EdgeTX/edgetx/assets/7026911/6e182f91-d639-43c8-94c4-22e1b7603bb1)


![image](https://github.com/EdgeTX/edgetx/assets/7026911/bd7838b9-a5b0-4bdd-a6c7-956a3b0ee1a5)


![image](https://github.com/EdgeTX/edgetx/assets/7026911/9ac8f809-f812-4ad1-baa2-65b6e50a51b9)

![image](https://github.com/EdgeTX/edgetx/assets/7026911/3520d520-8c65-4eba-be2b-4689b77757ee)

![image](https://github.com/EdgeTX/edgetx/assets/7026911/2126ffc9-0357-4d69-9b6a-3c971283d99e)

**Comment by jmxp69** [comment]
It might be worth considering an expo field for each rate. i.e. On 3D planes, high rates are usually accompanied by some high expo, like 70 where lower rates are usually more in the usual range 20-40. I look forward to trying this out. Great work!

**Comment by offer-shmuely** [comment]
good idea.

![image](https://github.com/EdgeTX/edgetx/assets/7026911/dbf3e95f-8e95-4d1a-b7e4-cca0cd8be29c)



![image](https://github.com/EdgeTX/edgetx/assets/7026911/a28b2cc2-6de1-4680-a83d-215fdeb8da42)

**Comment by offer-shmuely** [comment]
I use Hotar's UI sample as a background :-)


![image](https://github.com/EdgeTX/edgetx/assets/7026911/ab32517a-2499-4e95-a0a2-b1d059f1437d)

![image](https://github.com/EdgeTX/edgetx/assets/7026911/2bff0c00-db4d-4baf-8ad9-118c3ed18856)

![image](https://github.com/EdgeTX/edgetx/assets/7026911/157cbd8a-f875-46d3-88f2-776ce09e53c5)

**Comment by offer-shmuely** [comment]
implemented in RP
https://github.com/EdgeTX/edgetx-sdcard/pull/127

**Comment by Sea-Kerman** [comment]
Something like this might beg the implementation of some sort of "logic switch groups" which each have their own L01-L0whatever
